### PR TITLE
fix: warn when a symbol name is truncated by an invalid byte

### LINKED ISSUE #1810: A bad byte inside an identifier silently indexes the symbol under a truncated name
Found while fixing #1797. Distinct failure mode, same input class, and **not** fixed by that issue's remedy.

## Symptom

#1797 covers the case where a strict decode raises and the whole file is abandoned. There is a second, quieter case where nothing raises at all: the symbol is indexed under a **wrong name**, with no error, no log line, and no indication anything went wrong.

Measured on `main` through `GraphUpdater.run()`:

```
lua  function al\xffpha() ...   -> ['proj.a.other', 'proj.a.pha']
py   def al\xffpha(): ...       -> ['proj.a.beta',  'proj.a.pha']
go   func Al\xffpha() ...       -> ['proj.a.other', 'proj.a.pha']
js   function al\xffpha() ...   -> ['proj.a.other', 'proj.a.pha']
java int al\xffpha() ...        -> ['proj.a.C.pha()', ...]
```

`alpha` is indexed as `pha`. The leading `al` is gone.

## Cause

This is upstream of the decode, so `errors="replace"` does not reach it. tree-sitter treats the invalid byte as a token boundary and the `name` node covers only the bytes *after* it. The extractor then decodes that shortened node perfectly successfully. There is no exception to catch.

The two modes are decided by where the bad bytes land relative to what the extractor decodes as a single node:

```
lua  function Greeter\xff\xff.greet(n)  -> raises  -> whole file dropped   (#1797)
lua  function Greeter.gre\xffet(n)      -> no raise -> 'proj.a.Greeter.gre'
lua  function al\xffpha()               -> no raise -> 'proj.a.pha'
```

## Why it matters

A wrong name is arguably worse than a missing one. `pha` is a plausible-looking node: callers of `alpha` resolve to nothing and it reads as dead code, while `pha` reads as a real definition nobody calls. Dead-code, impact and test-reachability queries all get a confidently wrong answer rather than an obviously incomplete one.

It is also invisible to the fuzz harness: `fuzz_parse_source` detects this class by catching raises, and this mode never raises.

## Reproducer

```python
from pathlib import Path
from codebase_rag.graph_updater import GraphUpdater
from codebase_rag.parser_loader import load_parsers
from evals.cgr_graph import _StatefulIngestor

parsers, queries = load_parsers()
root = Path(tmpdir) / "proj"; root.mkdir()
(root / "a.lua").write_bytes(b"function al\xffpha() return 1 end\nfunction other() return 2 end\n")
store = _StatefulIngestor()
GraphUpdater(ingestor=store, repo_path=root, parsers=parsers,
             queries=queries, project_name="proj").run()
# -> {'proj.a.other', 'proj.a.pha'}; expected 'proj.a.alpha', not 'proj.a.pha'
```

## Possible directions

Not obvious which is right, hence filing rather than fixing:

1. Validate file bytes as UTF-8 once at read time and record a property on the Module node when they are not, so consumers can tell a mangled index from a clean one. Costs a scan per file.
2. Sanity-check an extracted name against its byte span: if the node's byte range is shorter than the identifier's extent in the source, mark it suspect.
3. Accept it, but make it visible: log at WARNING when a parsed file contains invalid UTF-8, so it is at least diagnosable.

Option 3 is the cheapest and would have made this findable without a fuzzer.

Separate from #1797 because the mechanism, the remedy and the detection are all different; #1797's `errors="replace"` fix does not change any of the behaviour above.

### PR DESCRIPTION
**Pull request by vitali87**
Closes #1810.

A bad byte inside an identifier is a token boundary to tree-sitter, so the `name` node covers only the bytes on one side of it. The extractor decodes that shortened node without error, and the symbol is indexed under a truncated name with nothing raised and nothing logged.

```
CLEAN FILE                     SAME FILE, ONE BAD BYTE
  billing.calculate_total        billing.calculate_t
  billing.main                   billing.main
```

## Scope: all 15 languages, measured

Same mutation applied to the repo's own per-language samples (`fuzz/build_corpus.py` SOURCES), driven through `GraphUpdater.run()`:

| outcome | count | languages |
|---|---|---|
| silently truncated | 13 | c, c_sharp, dart, go, java, javascript, php, python, rust, scala, sql, tsx, typescript |
| definition dropped | 2 | cpp, lua |

None survive intact. C# and Go truncate the containing **class**, taking every symbol under it: `App.Greeter` becomes `App.eeter`.

The instrument was validated before the table was believed: all 15 deterministic across repeated runs, and the mutation asserted to apply in each.

## The damage is not confined to the corrupt file

With `calculate_total` damaged in `lib.py`:

```
functions lost:      proj.lib.calculate_total
functions invented:  proj.lib.calculate_t
edges lost that originate in CLEAN files: 2
    proj.app.checkout   CALLS  proj.lib.calculate_total
    proj.report.summary CALLS  proj.lib.calculate_total
```

`app.py` and `report.py` are byte-perfect and both lose their call edges, so the function reads as dead code while a symbol nobody calls appears beside it. A wrong entry is worse than a missing one.

## Why the check is keyed on adjacency

Two narrower-sounding designs are both wrong, and the first shipped in this branch before the local reviewer caught it:

* **The definition node's span** contains the bad byte, but also spans the whole body, so it warns about correctly-indexed symbols whenever a bad byte sits in a body comment, docstring or string literal, and once per enclosing definition (one bad byte in a nested class produced four warnings).
* **The name node's own bytes** never contain the bad byte -- the grammar excludes it -- so `b'pha'` decodes cleanly in the corrupt and the clean case alike, detecting nothing at all.

What discriminates is the byte **immediately adjacent** to the name's span: the one the grammar split on.

`_name_span` locates the identifier by matching the extracted name rather than trusting `child_by_field_name`, because C nests it under `function_declarator` -- the field lookup alone silently lost c and sql.

## Base rate: why not a per-file check

Surveyed 105,352 real source files across six independent corpora:

| corpus | files | invalid UTF-8 |
|---|---:|---:|
| Rust crates | 16,855 | 0 |
| Go modules | 6,618 | 0 |
| macOS SDK | 11,575 | **136** |
| Homebrew | 38,837 | 26 |
| Python site-packages | 30,222 | 3 |
| this repo | 1,245 | 0 |
| **total** | **105,352** | **165** |

**None of the 165 corrupts a symbol.** Every one is latin-1 punctuation in a copyright header or an author's name (`©`, `ä`, `é`), and one is a `£` inside a regex literal -- code, not a comment. Five were indexed directly and all produced identical symbol sets to a repaired copy.

So a per-file signal would have fired 165 times and been wrong every time, at 1.2% of the macOS SDK. This check fires on none of them.

Cost is 0.059 microseconds per symbol (6ms for a 100,000-symbol repo); the bytes are already in hand, so no extra reads.

## Tests

13 in `test_truncated_symbol_names.py`: detection per truncating language, non-ASCII identifier controls (`café`, `élan`, `函数` -- a naive "is the name ASCII" check passes every test one would think to write and then fires on every non-English codebase), the false-alarm positions, and the nested-class case.

Reverting the helper to the previous implementation reddens exactly the two tests written to catch it. Two further mutations (guard disabled, name's own bytes) redden only the detection tests.

`test_the_name_nodes_own_bytes_cannot_detect_this` and `test_the_containment_oracle_cannot_detect_this` pin the two wrong designs as executable notes, because both are what a future reader will try to simplify this into. The containment oracle in particular is satisfied BY the corruption: `pha` really is a substring of `al\xffpha`.

String-literal and docstring positions are excluded from the false-alarm parametrisation with a note: they make the call pass raise `UnicodeDecodeError`, which is the separate open defect #1797 (fix in flight as #1812), and conftest's per-file-pass guard fails the test before this check is reached. Verified by driving the helper directly that both are handled.

cpp and lua are deliberately not covered: they invent no wrong-named symbol, and a missing definition is the honest, visible failure.

## Fuzz harness

`fuzz_parse_source` detected this class by catching `UnicodeDecodeError`, so it was structurally blind to the mode that raises nothing. It now checks the same property, with two reproducer seeds exempt from `test_parse_harness_runs_every_seed` and asserted the other way round in `test_known_defect_seeds_are_still_detected`, so an exemption cannot outlive the defect. `edge_bad_byte_between_defs.bin` is committed as the negative case pinning the scope limit.

## Verification

- 1,237 passed across truncated/fuzz/ingest/class_ingest/function/parser/language_spec, `-n 4` under the host lock.
- `ty check`: 9 diagnostics, identical to unmodified `main`.
- ruff check/format clean.

## Summary by CodeRabbit

* **Bug Fixes**
  * Added warnings when invalid UTF-8 bytes cause indexed symbol names to be truncated.
  * Improved detection across supported languages, symbol patterns, and synthesized names.
  * Prevented false warnings for valid non-ASCII names and unrelated invalid bytes.
  * Improved visibility into truncated names that can prevent caller resolution.

* **Tests**
  * Added coverage for truncated identifiers, Unicode names, parser edge cases, and malformed UTF-8 input.
  * Expanded fuzzing coverage for known parser defects and bounded defect reporting.

### CODE REVIEW

**Comment by vitali87** [comment]
@greptileai review

**Comment by vitali87** [comment]
All four review findings fixed in `c19a176b`. Each was reproduced here before being fixed, and each has a test that reddens when only that fix is reverted.

**Greptile P2 — valid UTF-8 warned (real, and the deepest of the four).** The adjacency probe decoded ONE byte either side, so `def alpha©()` under error recovery warned about well-formed source: `©` is c2 a9 and neither half decodes alone. The probe is now a 4-byte window (the longest UTF-8 sequence) with the partial character at the cut edge trimmed first. This was a genuine flaw in the design I described two commits ago, not just its implementation.

**Greptile P1 — C# destructors skipped detection.** Confirmed: `_name_span(destructor, ...)` returned `None` because ingestion synthesizes `~Greeter` while the source leaf is bare. A leading sigil is stripped before the lookup. One correction to the finding's detail: the C# grammar keeps the LEADING fragment, so the ingested name is `~Gr`, not `~eeter` — the span now resolves and the warning fires.

**Greptile P1 — fuzz oracle false crash.** Correct, and my omission: production moved to adjacency and the harness did not, so a bad byte in a body comment raised a false crash. Both now apply the same rule.

**CodeRabbit — field-bound names missed.** Correct on the mechanism. A TS class-field arrow is named by its parent field definition, so the identifier sits ABOVE the node handed to the check: `Greeter.eet` reached the graph unreported. `_name_span` now also walks out a bounded distance, which fixes the class rather than the one shape. JS is unaffected — verified it does not index class-field arrows as methods at all (the clean control indexes only `Greeter`), so there is no truncated name to report there.

**Sonar S5778.** Fixed: `seed.read_bytes()` moved out of the `pytest.raises` block, where a missing seed would have raised and been credited to the detector.

Unchanged after all four: detection covers 13 of 15 languages (cpp and lua drop the definition rather than mis-name it), all four false-alarm positions stay silent, and the five real invalid-UTF-8 files still produce zero warnings.

Verification: 1,240 passed across truncated/fuzz/ingest/class_ingest/function/parser/language_spec, `-n 4` under the host lock. `ty check` 9 diagnostics, identical to `main`.

@greptileai review

**Comment by vitali87** [comment]
Closed the non-blocking fuzz-coverage gap in `0faa05e3`.

The oracle now applies the same synthesized-name strip production does, so a regression in the C# destructor path is catchable by fuzzing rather than only by the unit test. Verified both directions: it raises on `~Gr\xffeeter` (reporting the ingested name `~Gr`) and stays silent on a clean `~Greeter`.

On the three notes about earlier findings being "resolved without explanation" — each was fixed in `c19a176b` and explained in the PR comment above it, then the threads were resolved because the repo ruleset requires resolution rather than replies to merge. Summarising here so the reasoning sits with the threads:

- **valid UTF-8 warning (P2)** — real, and the deepest of the three. The probe decoded one byte either side, so `def alpha©()` warned about well-formed source: `©` is c2 a9 and neither half decodes alone. Now a 4-byte window with the partial character at the cut edge trimmed.
- **C# destructors (P1)** — confirmed `_name_span` returned `None`; a leading sigil is now stripped first. One correction to the finding's detail: the grammar keeps the LEADING fragment, so the ingested name is `~Gr`, not `~eeter`.
- **fuzz oracle body bytes (P1)** — my omission; production had moved to adjacency and the harness had not. Both now apply the same rule.

@greptileai review

**Comment by vitali87** [comment]
`Fuzz (address)` failed on the previous head and is fixed in `a6925ff1`.

The fuzzer found `sa\x92O()\x01` — in C# the bad byte splits the token and `O` is indexed under a truncated name. I reproduced it locally: **the oracle was correct**, that input really does exhibit #1810.

Halting is the wrong response for a defect production now DETECTS and reports. libFuzzer invents a mangled identifier within seconds, so every run would stop on the first one and hide every other defect behind it — the fuzz target would effectively test one thing. The file already applies exactly this reasoning to #1797, which it suppresses with a "remove when fixed" note.

The oracle now records to `_TRUNCATED_NAMES` and continues, and stays load-bearing because `test_known_defect_seeds_are_still_detected` requires both reproducers to trip it. Verified by silently disabling the append: 4 tests redden, so this cannot decay into an oracle that does nothing.

The seed exemption from `test_parse_harness_runs_every_seed` is removed with it — nothing raises now, so all 25 seeds run under the same assertion rather than two being skipped.

Verified: the CI crash input no longer raises but is recorded (`[(CSHARP, 'O')]`), all 25 corpus seeds run clean, and both reproducers still record their truncated names (`'pha'`, `'alph'`). 1,240 tests pass.

@greptileai review

**Comment by vitali87** [comment]
Fixed in `eda16a44`. Real problem, reproduced before fixing: 500 identical inputs produced 500 entries, and libFuzzer runs millions of iterations, so the collection would have exhausted memory and killed the run — the same "the harness crashed itself" failure that recording rather than raising exists to avoid.

Now a set with a cap of 128. 5,000 inputs give 1 entry rather than 5,000.

Worth flagging that my first test for this was decoration: it fed the SAME input repeatedly, so the set deduplicated it to one entry and the test passed with or without the cap. Caught by mutating the cap away and seeing nothing redden. It now uses a distinct identifier per iteration, and removing the cap reddens exactly that test.

1,241 tests pass.

@greptileai review

File: fuzz/fuzz_parse_source.py [hunk 1]
```diff
@@ -81,17 +90,125 @@ def _extract_names(language: cs.SupportedLanguage, root: object) -> None:
...
+            # known, reported behaviour rather than a crash: halting the
+            # fuzzer on it would stop the run on the first mangled identifier
+            # libFuzzer invents and hide every other defect behind it.
+            #
+            # The oracle is still load-bearing. It is asserted the other way
+            # round in `test_known_defect_seeds_are_still_detected`, which
+            # requires the two corpus reproducers to trip it, so a regression
+            # that stops detecting truncation reddens there. Raise here again
+            # once the extractors stop truncating rather than warning.
+            _TRUNCATED_NAMES.append((language, name))
```
[the remark this answers is not in the document]
Comment (vitali87) [reply fuzz/fuzz_parse_source.py]: Already fixed — this review is anchored to `a6925ff1`, the head before the fix.

At the current head `eda16a44` it is a bounded set, not an unbounded list:

```python
_TRUNCATED_NAME_LIMIT = 128
_TRUNCATED_NAMES: set[tuple[cs.SupportedLanguage, str]] = set()
```

Same finding CodeRabbit raised independently; I reproduced it (500 identical inputs produced 500 entries) and fixed it in `eda16a44`. 5,000 inputs now give 1 entry.

Your 25,000-distinct-names measurement is the better repro, and it is what the regression test now models: my first version of that test fed the SAME input repeatedly, so the set deduplicated it and the test passed with or without the cap. It now uses a distinct identifier per iteration, and removing the cap reddens it.

**Comment by vitali87** [comment]
Rebased onto `main` now that #1812 has merged, and the rebase improved the result.

`decode_node_text` changed what two languages do with a bad byte inside an identifier:

- **cpp** now truncates where it used to drop the definition, so the adjacency check covers it with no code change — detection went 13 → 14 of 15.
- **lua** exposed a **second damage shape**. It keeps the bad byte in an ERROR node BETWEEN the two identifiers of a dotted name, so the whole expression decodes to `Greeter.gr\ufffdeet`. Nothing is truncated, so nothing is adjacent to a shortened span and the adjacency test finds nothing — the name itself carries the damage.

U+FFFD in an extracted name is only ever there because a decode put it there (no legitimate identifier contains it, in any language), so that signal needs no adjacency test and cannot false-alarm on valid source. Removing it reddens exactly the new Lua test.

**Detection is now 15 of 15 languages**, with all four false-alarm positions still silent and zero warnings on the five real invalid-UTF-8 files.

Two conflicts in the rebase, both resolved by hand: `main` removed the #1797 `UnicodeDecodeError` suppression in the same block where I had added the #1810 check. Kept both — no suppression, plus the check. Verified nothing was lost by comparing the symbol inventory against the pre-rebase commit (8 and 65 definitions, unchanged) and confirming `decode_node_text` (13 references) and `warn_if_name_truncated` (all three sites) both survive.

Also verified Dart independently, per the #1812 report: `class Gree\xff\xffter` indexes as `proj.a.Gree` and is reported.

Verification: 1,592 passed, 4 skipped, 0 failed (`-n 4` under the host lock, including dart and lua). `ty check` 9 diagnostics, identical to `main`.

@greptileai review

**Comment by vitali87** [comment]
Fixed in `c517634d`. Valid finding, reproduced before fixing: the oracle recorded nothing for `function Greeter.gr\xffeet(n)` while production reported it.

This is the **second time** production and the harness have drifted in this PR, both in the same direction — production gains a rule and the oracle keeps the old one. First the synthesized-name strip, now the replacement-character shape. Worth naming as a pattern rather than two incidents: any rule added to `warn_if_name_truncated` has to be mirrored in `_check_truncated_name`, because the harness cannot be tested by the production tests and vice versa.

Verified both directions: the oracle now records `Greeter.gr\ufffdeet` for the malformed Lua source and stays silent on the clean equivalent. Removing the rule reddens exactly the new test.

On the four threads noted as "resolved without an explanatory reply": each was fixed in `c19a176b`, explained in the PR comment posted alongside it, and then resolved because the ruleset requires resolution rather than replies to merge. The explanations are in the timeline rather than in the threads themselves — I have since been replying in-thread before resolving, which is the better habit.

Verification: 1,247 passed (`-n 4` under the host lock). `ty check` 9 diagnostics, identical to `main`.

@greptileai review
